Sources/Sentry/Public/SentryOptions.h

```diff
@@ -199,7 +199,6 @@ NS_SWIFT_NAME(Options)
  */
 @property (nonatomic) SentryScope * (^initialScope)(SentryScope *);
 
-#if SENTRY_HAS_UIKIT
```

Comment:
Oh hm, I thought I had added it in the middle of this PR and then decided not to do it here. Guess I was wrong 😆 

But, now that `SENTRY_HAS_UIKIT` has been extended to include whether a `GCC_PREPROCESSOR_DEFINITIONS` is set, I think we can't use this in headers any longer, since the header could be included into targets that don't get our `GCC_PREPROCESSOR_DEFINITIONS` values, this would never evaluate to true.

What I did instead was override the setters for these options as no-ops for `!SENTRY_HAS_UIKIT` and print a warning log that they do nothing for configurations that don't link UIKit.

Reply:
I'll reevaluate this claim. I was unable to build the sample apps because the options gated by this were not able to be located. But SentryDefines.h itself is also a public header, and seemed to evaluate the macros correctly.

Reply:
So there's something going on with how regions surrounded by `SENTRY_HAS_UIKIT` are visible to Swift code that imports the Sentry module. The last thing that's holding me up is that when I remove gated compilation here: https://github.com/getsentry/sentry-cocoa/blob/2fe817d73be7b83f45f72bd55adf132d0f594bcb/Samples/iOS-Swift/iOS-Swift/ExtraViewController.swift#L34-L38

the iOS-Swift build fails because the compiler can't see `PrivateSentrySDKOnly.currentScreenFrames`. The stranger thing now is that even if I remove the `#if SENTRY_HAS_UIKIT` surrounding that and other declarations in PrivateSentrySDKOnly.h, it _still_ isn't building. I'm digging in a little more but no idea why this is happening right now :/ 

This is essentially the same thing I saw in the other apps though, which is why I removed those, and those builds now work. I'm wondering if there's something special about PrivateSentrySDKOnly.h though, since it's in a different subdirectory.

Reply:
Yeah there's definitely something going on with Swift apps consuming Sentry being unable to use declarations from Sentry.framework/PrivateHeaders, I also tried it with another one and that fails too: 
![image](https://github.com/getsentry/sentry-cocoa/assets/3241469/91c00f5b-cc55-4dcf-b71b-b067fda613ba)

Reply:
So the issue was that the GCC_PREPROCESSOR_DEFINITION setting being used to set the macro definition in SentryDefines.h had no effect in headers delivered in the framework bundle. So for headers exposed in the framework bundle, I reverted to the previous type of macro, under a new name `SENTRY_UIKIT_AVAILABLE`, and stubbed the implementations depending on `SENTRY_HAS_UIKIT` (which, again, means it's available _and_ linked). You can see https://github.com/getsentry/sentry-cocoa/pull/3175/commits/ccd24659d579657049fd56f63822fb5988329c02

Reply:
Thanks for explaining, but why do we now use a combination of both `SENTRY_UIKIT_AVAILABLE` and `SENTRY_HAS_UIKIT` https://github.com/getsentry/sentry-cocoa/blob/a697fca629f46b6296f29bf28ad4d303c7eecbef/Sources/Sentry/SentryOptions.m#L635-L645

Wouldn't only using `SENTRY_HAS_UIKIT` be sufficient? I don't understand why we want to expose functionality when UIKit is available on the platform but not linked and to provide a fallback that prints a warning. Only using `SENTRY_HAS_UIKIT` would reduce complexity, or is that not possible?

Reply:
The old way that `SENTRY_HAS_UIKIT` was defined only tells us if we're compiling for either an iOS or tvOS target. So this works to compile out UIKit code for eg macOS, but if we want to compile out UIKit code on iOS/tvOS, we need another option to switch over. The only way I thought of to do so was with the new build configs that deliver a `GCC_PREPROCESSOR_DEFINITION` that I called `SENTRY_UIKIT_LINKED`, which we now check for when defining `SENTRY_HAS_UIKIT`. This is why I renamed what used to be `SENTRY_HAS_UIKIT` (`TARGET_OS_IOS || TARGET_OS_TVOS`) to `SENTRY_UIKIT_AVAILABLE` and changed `SENTRY_HAS_UIKIT` to be not only `SENTRY_UIKIT_AVAILABLE`, but also `SENTRY_UIKIT_LINKED` for the build configs that do so. I think the changed naming is more accurate for both cases as well. Let me know if this makes sense to you; I also wrote this up in the develop-docs/README.md in https://github.com/getsentry/sentry-cocoa/pull/3175/commits/2f5c366c2f960f04d96d2df2e0cc4282f53005de

Reply:
I agree with Philipp, it seems to me that `#if SENTRY_HAS_UIKIT` is what we should use, because it addresses both situation, it's the right target and linking UIKit is enabled.

Reply:
> What I saw was our test apps would not compile for both Debug and Debug_without_UIKit, they'd break on the latter because the symbol was compiled out but the sample app doesn't inherit the GCC_PREPROCESSOR_DEFINITION.

Not compiling for Debug is suspicious. Our sample apps not compiling for `Debug_without_UIKit`  should be fine. I guess we need a new sample without UIKit to validate `Debug_without_UIKit`.

> We also don't want to make people have to write #if/#endif regions around code we conditionally compile. We figured it was better to not break their builds and just print warnings (and I added documentation to all such public API).

We absolutely can't break existing builds in a minor version and writing `#if/#endif regions` is also suboptimal.

I agree with @brustolin. @armcknight, can we somehow make it work without breaking customers? 

As you said, maybe it's better to discuss this over a call instead of ping-pong here on GH.

Reply:
>> our test apps would not compile for both Debug and Debug_without_UIKit

> Not compiling for Debug is suspicious. Our sample apps not compiling for Debug_without_UIKit should be fine

Sorry, I should've been clearer. This is actually the case: before, it built for debug, but not debug_without_uikit. After my changes, it built for both, and only emitted the warning logs in debug_without_uikit. I did this so no code changes would be necessary in the sample app. This is my definition of not breaking customers.

I can add assertions instead of just warning logs, so that debug builds crash, but not production. How does that sound? I will push a commit with this change.